### ISSUE TITLE
Get rid of persistent build errors due to non-atomic file writes in GHC and stack

### ISSUE THREAD

**Issue by nh2** [commercialhaskell/stack#4559]
GHC as of writing does not ensure that files are written atomically.

This means that a Ctrl+C, kill or reboot at the right time can result in truncated files.
GHC does not detect this, so that resuming/rerunning the build with `ghc --make` continues to show up error messages.

In such situation, the only workaround is to wipe all files (e.g. `stack`/`cabal`/`make` `clean`).

Specifically, we've observed the following to happen:

* object files being written half-way, resulting in persistent linker errors
* executable files being written half-way, so they can start executing but then crash
* users reporting downstream tooling bug reports that only a build directory wipe helped

There may be situations where `stack` has this problem too, but so far we believe all occurrences of this that we see are in GHC.

## Repro

@lehins made a repro script that shows the problem in GHC at https://github.com/lehins/exec-kill-loop

## Planned solution

* GHC should use atomic writes (write to temp file, then `rename()` syscall).
  * [x] Make repro script that shows the problem ([done](https://github.com/lehins/exec-kill-loop))
  * [ ] Decrease repro waiting time from hours to a few minutes by killing just at the right time, doing syscall interception
  * [ ] Find all places where GHC does non-atomic file writes, and fix them
  * [ ] If possible, add this as a system test to GHC CI

### Issues for which this _may_ be the reason

* https://github.com/commercialhaskell/stack/issues/2422#issuecomment-460362698
* https://github.com/commercialhaskell/stack/issues/4399#issuecomment-460362870

**Comment by nh2** [comment]
In case we need to add atomic writes for stack, in https://github.com/commercialhaskell/rio/pull/138 @roman introduced atomic+durable file writes that we can use.

However, as per [this](https://github.com/commercialhaskell/rio/pull/138/files#diff-56a7ad26dec349be191aa4b8474d47e8R41):

```haskell
There is no function exported by this module that provides /only/ atomicity.
```

We probably want to add that since for many cases, atomic-but-not-durable is enough, and durability (`fsync`) makes things a lot slower.

**Comment by nh2** [comment]
New task done:

> * [x] Decrease repro waiting time from hours to a few minutes by killing just at the right time, doing syscall interception

Here comes [hatrace](https://github.com/nh2/hatrace), a Haskell-API for `strace`-style syscall incerception and scripting.

I made it for my work on https://phabricator.haskell.org/D42, but it's very useful for this problem, too.

**Comment by nh2** [comment]
My GHC patch for writing `.o` files atomically got merged:

https://gitlab.haskell.org/ghc/ghc/merge_requests/391

This should fix the biggest source for these errors, but the other files GHC and its subprograms write need to be done.

**Comment by snoyberg** [comment]
@nh2 I don't think this issue belongs on the Stack issue tracker, as it's entirely upstream. Any objection to closing it?

**Comment by nh2** [comment]
The main GHC bughas been fixed, but I still wanted to do an investigation with [hatrace](https://github.com/nh2/hatrace#readme) on

> There may be situations where `stack` has this problem too

to ensure that there's no case where stack may write non-atomic files.

**Comment by nh2** [comment]
In https://github.com/nh2/hatrace/pull/9 @qrilka implemented a mode for `hatrace` that can point out non-atomic writes happening.

In https://github.com/nh2/hatrace/pull/9#issuecomment-477600387 we recoded its output for a `stack build` invocation. It yields some paths that are written non-atomically, and we should check whether those are benign or should be improved.

Especially interesting are the entries

```
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/install/x86_64-linux/lts-8.22/8.0.2/lib/x86_64-linux-ghc-8.0.2/call-haskell-from-anything-1.1.0.0-6Xcajdv7jFLAvwMt571tKe/st4joiWL/DataKinds.o"
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/install/x86_64-linux/lts-8.22/8.0.2/lib/x86_64-linux-ghc-8.0.2/call-haskell-from-anything-1.1.0.0-6Xcajdv7jFLAvwMt571tKe/st4joiWL/Msgpack.o"
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/install/x86_64-linux/lts-8.22/8.0.2/lib/x86_64-linux-ghc-8.0.2/call-haskell-from-anything-1.1.0.0-6Xcajdv7jFLAvwMt571tKe/st4joiWL/TH.o"
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/install/x86_64-linux/lts-8.22/8.0.2/lib/x86_64-linux-ghc-8.0.2/call-haskell-from-anything-1.1.0.0-6Xcajdv7jFLAvwMt571tKe/st4joiWL/TypeUncurry.o"
```

which disappear after `stack build` completes.

**Comment by nh2** [comment]
I think the fastest way to investigat it would be to add a command like `hatrace --kill-after-write /path/to/Myfile.o`, and then we can run it on each file shown by `--find-nonatomic-writes`, and afterwards run another `stack build` or `stack repl` and see if it recovers or crashes.

**Comment by snoyberg** [comment]
This is referring to writes in GHC itself, not Stack, correct?

**Comment by qrilka** [comment]
@snoyberg the most of non-atomically written files are form GHC, I think we see some from Cabal e.g. from the `copy` operation but also there are some produced by Stack, I think at least these:
```
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/dist/x86_64-linux/Cabal-1.24.2.0/stack-config-cache"
 - "/home/niklas/src/haskell/call-haskell-from-anything/.stack-work/install/x86_64-linux/lts-8.22/8.0.2/flag-cache/call-haskell-from-anything-1.1.0.0-6Xcajdv7jFLAvwMt571tKe"
```
And during our call Niklas was able to reproduce Stack failure by cutting one of those

**Comment by snoyberg** [comment]
Awesome, that's good to know @qrilka. I believe both of those are going away with @borsboom's changes to remove `store`.

**Comment by qrilka** [comment]
We were testing 1.9.3 so it makes total sense to do proper testing with `master`

**Comment by borsboom** [comment]
> Awesome, that's good to know @qrilka. I believe both of those are going away with @borsboom's changes to remove store.

That is correct.

**Comment by snoyberg** [comment]
Possibly relevant?

* https://github.com/commercialhaskell/stack/issues/3248
* https://github.com/commercialhaskell/stack/issues/3599

**Comment by nh2** [comment]
Yes, those look relevant.

The last one suffers from non-atomic writes to `.hi` files, which is something I haven't fixed yet in GHC (only object files).
